docs/notebooks/code_overview.pct.py

```diff
@@ -0,0 +1,238 @@
+# %% [markdown]
+# # An overview of Trieste types
+
+# %% [markdown]
+# Trieste is dedicated to Bayesian optimization, the process of finding the *optimal values of an expensive, black-box objective function by employing probabilistic models over observations*. This notebook explains how the different parts of this process are represented by different types in the code, and how these types can be extended.
+
+# %% [markdown]
+# ## Key types
+#
+# The following types represent the key concepts in Trieste. For a full listing of all the types in Trieste, see the API reference.
+
+# %% [markdown]
+# ### `Observer`
+#
+# The `Observer` type definition represents the black box objective function. Observers are functions that accept query points and return observations. Observations are either a single objective value that we wish to optimize, or a dictionary of mutiple tagged values that must be combined somehow, for example an objective and an inequality constraint. Objective values can be either single or multi-dimensional (see [multi-objective optimization](multi_objective_ehvi.ipynb)).
+
+# %% [markdown]
+# ### `Dataset`
+#
+# The `Dataset` container class represents the query points and observations from a single observer. Observers with multiple observations are represented by a dictionary of multiple tagged `Dataset`s.
+
+# %% [markdown]
+# ### `ProbabilisticModel`
+#
+# The `ProbabilisticModel` protocol represents any probabilistic model used to model observations. Like for `Dataset`, observers with multiple observations are modelled by a dictionary o multiple tagged models.
+#
+# At it simplest, a `ProbabilisticModel` is anything that implements a `predict` and `sample` method. However, many algorithms in Trieste depend on models with additional features, which are represented by the various subclasses of `ProbabilisticModel`. The standard Bayesian optimizer uses `TrainableProbabilisticModel` models, which also implement `update` and `optimize` methods. Specific acuqisition functions may require other features, represented by classes like `SupportsPredictJoint` and `SupportsGetObservationNoise`. Since these are defined as protocols, it is possible to define and depend on the intersections of different model types (e.g. only support models that are both `SupportsPredictJoint` and `SupportsGetObservationNoise`).
+#
+# Multiple models can also be combined into a single `ModelStack` model that combines their outputs for prediction and sampling. This can be useful when modelling multi-objective observations with independent models. There are also constructors like `TrainableModelStack` and `PredictJointModelStack` that combine specific types of model for use with code that requires that type, delegating the relevant methods where appropriate.
```

Comment:
Would start with a slightly more general sentence about multiple outputs perhaps? e.g. something along the lines of

There are multiple ways to deal with multiple outputs: one can use a single multi-output model or a dictionary of models.

Maybe also add that `ModelStack` essentially emulates a single multi-output model from independent single-output models? and say it's useful for acquisitions that expect a multioutput model such as our multi-objective ones?